File: src/db-toolkit/operations/analytics/pdf_export.py
"""PDF export for analytics metrics."""

from typing import Dict, List, Any
from datetime import datetime

from db_toolkit.utils.pdf import PDFReport, format_bytes, format_duration, truncate_text


def generate_analytics_pdf(
    connection_name: str,
    metrics: Dict[str, Any],
    historical_data: List[Dict[str, Any]],
    slow_queries: List[Dict[str, Any]],
    table_stats: List[Dict[str, Any]]
) -> bytes:
    """Generate PDF report for analytics data."""
    
    pdf = PDFReport("Database Analytics Report")
    pdf.add_title(subtitle=f"Connection: {connection_name}")
    
    # Current Metrics
    pdf.add_section("Current Metrics")
    system_stats = metrics.get('system_stats', {})
    pdf.add_metric_grid([
        {"label": "Active Connections", "value": str(metrics.get('active_connections', 0)), "color": "#10b981"},
        {"label": "Idle Connections", "value": str(metrics.get('idle_connections', 0)), "color": "#6b7280"},
        {"label": "Database Size", "value": format_bytes(metrics.get('database_size', 0)), "color": "#3b82f6"},
        {"label": "CPU Usage", "value": f"{system_stats.get('cpu_usage', 0):.1f}%", "color": "#f59e0b"},
        {"label": "Memory Usage", "value": f"{system_stats.get('memory_usage', 0):.1f}%", "color": "#8b5cf6"},
        {"label": "Disk Usage", "value": f"{system_stats.get('disk_usage', 0):.1f}%", "color": "#ec4899"},
    ])
    
    # Query Statistics
    if metrics.get('query_stats'):
        pdf.add_section("Query Distribution")
        query_data = [[qtype, str(count)] for qtype, count in metrics['query_stats'].items()]
        pdf.add_table(query_data, headers=['Query Type', 'Count'])
    
    # Slow Queries
    if slow_queries:
        pdf.add_page_break()
        pdf.add_section("Slow Query Log (Last 24 Hours)")
        slow_data = [
            [
                q['timestamp'][:19],
                format_duration(q['duration']),
                truncate_text(q['user'], 20),
                truncate_text(q['query'], 80)
            ]
            for q in slow_queries[:20]
        ]
        pdf.add_table(slow_data, headers=['Timestamp', 'Duration', 'User', 'Query'])
    
    # Table Statistics
    if table_stats:
        pdf.add_page_break()
        pdf.add_section("Table Statistics (Top 20)")
        table_data = []
        for t in table_stats[:20]:
            table_name = t.get('tablename') or t.get('table_name') or t.get('collection', 'N/A')
            size = t.get('size', 'N/A')
            if isinstance(size, (int, float)):
                size = format_bytes(int(size))
            rows = t.get('row_count', t.get('n_live_tup', 0))
            table_data.append([truncate_text(table_name, 40), str(size), str(rows)])
        pdf.add_table(table_data, headers=['Table', 'Size', 'Rows'])
    
    return pdf.build()

File: src/db-toolkit-ui/electron/backend/operations/analytics/pdf-export.js
/**
 * PDF export for analytics metrics.
 */

const PDFDocument = require('pdfkit');

function generateAnalyticsPDF(connectionName, metrics, historicalData, slowQueries, tableStats) {
  const doc = new PDFDocument({ margin: 50 });
  const chunks = [];
  
  doc.on('data', chunk => chunks.push(chunk));
  
  // Title with green theme
  doc.fontSize(24).fillColor('#10b981').text('Database Analytics Report', { align: 'center' });
  doc.moveDown(0.5);
  doc.fontSize(14).fillColor('#000').text(`Connection: ${connectionName}`, { align: 'center' });
  doc.moveDown(0.5);
  doc.fontSize(9).fillColor('#6b7280').text(`Generated: ${new Date().toLocaleString()}`, { align: 'center' });
  doc.moveDown(2);
  
  // Current Metrics
  doc.fontSize(16).fillColor('#10b981').text('Current Metrics');
  doc.moveDown(0.5);
  
  const activeConn = metrics.active_connections || 0;
  const idleConn = metrics.idle_connections || 0;
  const dbSize = formatBytes(metrics.database_size || 0);
  
  doc.fontSize(10).fillColor('#000');
  doc.text(`Active Connections: ${activeConn}`);
  doc.text(`Idle Connections: ${idleConn}`);
  doc.text(`Database Size: ${dbSize}`);
  doc.moveDown();
  
  // Query Stats
  const queryStats = metrics.query_stats || {};
  if (Object.values(queryStats).some(v => v > 0)) {
    doc.fontSize(16).fillColor('#10b981').text('Query Distribution');
    doc.moveDown(0.5);
    doc.fontSize(10).fillColor('#000');
    for (const [type, count] of Object.entries(queryStats)) {
      if (count > 0) doc.text(`${type}: ${count}`);
    }
    doc.moveDown();
  }
  
  // Historical Data
  if (historicalData && historicalData.length > 0) {
    doc.fontSize(16).fillColor('#10b981').text('Historical Trends');
    doc.moveDown(0.5);
    doc.fontSize(10).fillColor('#000');
    doc.text(`Data points collected: ${historicalData.length}`);
    const latest = historicalData[historicalData.length - 1];
    if (latest) {
      doc.text(`Latest snapshot: ${latest.timestamp || 'N/A'}`);
      doc.text(`Active connections: ${latest.active_connections || 0}`);
      doc.text(`Database size: ${formatBytes(latest.database_size || 0)}`);
    }
    doc.moveDown();
  }
  
  // Current Queries
  const currentQueries = metrics.current_queries || [];
  if (currentQueries.length > 0) {
    doc.fontSize(16).fillColor('#10b981').text('Current Active Queries');
    doc.moveDown(0.5);
    doc.fontSize(9).fillColor('#000');
    for (const q of currentQueries.slice(0, 10)) {
      doc.text(`PID: ${q.pid} | User: ${q.usename} | Duration: ${q.duration}s`);
      doc.text(`Query: ${truncateText(q.query, 80)}`);
      doc.moveDown(0.3);
    }
    doc.moveDown();
  }
  
  // Slow Queries
  if (slowQueries && slowQueries.length > 0) {
    doc.addPage();
    doc.fontSize(16).fillColor('#10b981').text('Slow Query Log (Last 24 Hours)');
    doc.moveDown(0.5);
    doc.fontSize(9).fillColor('#000');
    for (const q of slowQueries.slice(0, 20)) {
      doc.text(`Time: ${q.timestamp?.slice(0, 19) || 'N/A'} | Duration: ${formatDuration(q.duration)}`);
      doc.text(`User: ${q.user} | Query: ${truncateText(q.query, 80)}`);
      doc.moveDown(0.3);
    }
  }
  
  // Table Stats
  if (tableStats && tableStats.length > 0) {
    doc.addPage();
    doc.fontSize(16).fillColor('#10b981').text('Table Statistics (Top 20)');
    doc.moveDown(0.5);
    doc.fontSize(9).fillColor('#000');
    for (const t of tableStats.slice(0, 20)) {
      const tableName = t.tablename || t.table_name || t.collection || 'N/A';
      const size = typeof t.size === 'number' ? formatBytes(t.size) : t.size || 'N/A';
      const rows = t.row_count || t.n_live_tup || 0;
      doc.text(`${truncateText(tableName, 40)} | Size: ${size} | Rows: ${rows}`);
    }
  }
  
  doc.end();
  
  return new Promise((resolve) => {
    doc.on('end', () => resolve(Buffer.concat(chunks)));
  });
}

function formatBytes(bytes) {
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  let value = bytes;
  let unitIndex = 0;
  while (value >= 1024 && unitIndex < units.length - 1) {
    value /= 1024;
    unitIndex++;
  }
  return `${value.toFixed(2)} ${units[unitIndex]}`;
}

function formatDuration(seconds) {
  if (seconds < 1) return `${(seconds * 1000).toFixed(0)}ms`;
  if (seconds < 60) return `${seconds.toFixed(2)}s`;
  if (seconds < 3600) return `${(seconds / 60).toFixed(1)}m`;
  return `${(seconds / 3600).toFixed(1)}h`;
}

function truncateText(text, maxLength = 100) {
  if (text.length <= maxLength) return text;
  return text.slice(0, maxLength - 3) + '...';
}

module.exports = { generateAnalyticsPDF };
